DataDog/datadog-operator
# Fix DatadogAgent secret backend usage

### What does this PR do?

The secret backend feature for DatadogAgent was not working properly. This fixes the feature so that any credentials can use it. In addition, the Operator no longer creates a secret if it's not needed (i.e. it will not create an empty secret).

Other changes:
- create secret_common.go to hold common functions
- add new examples/datadogagent/datadog-agent-secret-backend.yaml
- update secret backed documentation

### Motivation

Fix

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Use the new `datadog-agent-secret-backend.yaml` example file to test the secret backend feature.

For completeness, can test the following cases (make sure the keys are read properly and the Agent can connect to the Cluster Agent):

(Create a secret containing credentials with `kubectl create secret generic test-secret --from-literal=api_key='<token>' --from-literal=app_key='<token>' --from-literal=token='<token>' `)

1. secret backend not used (check that Operator creates a secret based on provided credentials)
2. only API key uses secret backend
3. only App key uses secret backend
4. only token uses secret backend
5. all three (API key, App key, token) use secret backend
6. set API key using APISecret
7. set App key using APPSecret
9. enable ExternalMetrics with keys configured in ExternalMetricsCredentials, no secret backend
10. enable ExternalMetrics with keys configured in ExternalMetricsCredentials, with secret backend
